src/compute-client/src/controller/instance.rs

```diff
@@ -2145,39 +2129,170 @@ impl<T: ComputeControllerTimestamp> CollectionState<T> {
+                    // an `initial_as_of`.
+                    soft_assert_or_log!(
+                        refresh_schedule.is_none(),
+                        "`refresh_schedule` without an `initial_as_of`: {collection_id}"
+                    );
+                    None
+                }
+            };
+        let frontiers = (!storage_sink).then(|| FrontiersIntrospectionState::new(as_of));
+
+        let self_ = Self {
+            collection_id,
+            introspection_tx,
+            frontiers,
+            refresh,
+        };
+
+        self_.report_initial_state();
+        self_
+    }
+
+    /// Reports the initial introspection state.
+    fn report_initial_state(&self) {
+        if let Some(frontiers) = &self.frontiers {
+            let row = frontiers.row_for_collection(self.collection_id);
+            let updates = vec![(row, 1)];
+            self.send(IntrospectionType::Frontiers, updates);
+        }
+
+        if let Some(refresh) = &self.refresh {
+            let row = refresh.row_for_collection(self.collection_id);
+            let updates = vec![(row, 1)];
+            self.send(IntrospectionType::ComputeMaterializedViewRefreshes, updates);
+        }
+    }
+
+    /// Observe the given current collection frontiers and update the introspection state as
+    /// necessary.
+    fn observe_frontiers(
+        &mut self,
+        read_frontier: AntichainRef<T>,
+        write_frontier: AntichainRef<T>,
+    ) {
+        self.update_frontier_introspection(read_frontier, write_frontier);
+        self.update_refresh_introspection(write_frontier);
+    }
+
+    fn update_frontier_introspection(
+        &mut self,
+        read_frontier: AntichainRef<T>,
+        write_frontier: AntichainRef<T>,
+    ) {
+        let Some(frontiers) = &mut self.frontiers else {
+            return;
+        };
+
+        if frontiers.read_frontier.borrow() == read_frontier
+            && frontiers.write_frontier.borrow() == write_frontier
+        {
+            return; // no change
+        };
+
+        let retraction = frontiers.row_for_collection(self.collection_id);
+        frontiers.update(read_frontier, write_frontier);
+        let insertion = frontiers.row_for_collection(self.collection_id);
+        let updates = vec![(retraction, -1), (insertion, 1)];
+        self.send(IntrospectionType::Frontiers, updates);
+    }
+
+    fn update_refresh_introspection(&mut self, write_frontier: AntichainRef<T>) {
+        let Some(refresh) = &mut self.refresh else {
+            return;
+        };
+
+        let retraction = refresh.row_for_collection(self.collection_id);
+        refresh.frontier_update(write_frontier);
+        let insertion = refresh.row_for_collection(self.collection_id);
+
+        if retraction == insertion {
+            return; // no change
+        }
+
+        let updates = vec![(retraction, -1), (insertion, 1)];
```

Comment:
If you want to make this a bit more generic, you could consolidate the vector and then check for emptiness. But the code as-is works too.

Reply:
Hm, I guess we could put that consolidate-and-check code into `send` so it applies to every batch of updates we send. At the moment, this is the only place where retraction and insertion can actually be the same, so moving the check to `send` doesn't actually reduce duplication while slightly reducing performance due to unnecessary consolidations, so I'll leave it as is for now.